kv: TestTxnNegotiateAndSendDoesNotBlock failed [failed to generate INITIAL snapshot: cannot read undeclared span /Local/.../u/RaftTruncatedState]

kv.TestTxnNegotiateAndSendDoesNotBlock [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11737675?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11737675?buildTab=artifacts#/) on master @ [0310d119569b59b9efe1b7bbc439c2c604217a4c](https://github.com/cockroachdb/cockroach/commits/0310d119569b59b9efe1b7bbc439c2c604217a4c):


```
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/logstore/stateloader.go:98 +0x39f
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver/stateloader.StateLoader.Load({{{0xc000937700?, 0x0?, 0x0?}}}, {0xb03dc20, 0xc008e54780}, {0x7f4e140aba58, 0xc007d7cd80}, 0xc00fd48626?)
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/stateloader/stateloader.go:89 +0x505
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.snapshot({_, _}, {0x4c, 0xca, 0x1e, 0x38, 0x53, 0x73, 0x49, 0xad, ...}, ...)
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raftstorage.go:385 +0x2b8
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).GetSnapshot(0xc001dc0d80, {0xb03dc20, 0xc008e544e0}, 0x0?, {0x4c, 0xca, 0x1e, 0x38, 0x53, 0x73, ...})
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raftstorage.go:289 +0xa1c
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).followerSendSnapshot(0xc001dc0d80, {0xb03dc20, 0xc008e544b0}, {0x4b0b4c?, 0x0?, 0x4b2637?, 0x0?}, 0xc0023b2680)
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:3118 +0x4fb
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleDelegatedSnapshot(0xc00b001500, {0xb03db78, 0xc0037cad70}, 0xc0023b2680)
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:186 +0x2f5
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).InternalDelegateRaftSnapshot(0xc005a82380, {0xb03db78, 0xc0037cad70}, 0xc0023b2680)
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:591 +0x2de
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*RaftTransport).DelegateRaftSnapshot(0xc005a82380, {0xb07c2d8, 0xc000a4e270})
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/raft_transport.go:554 +0x15d
          |   | github.com/cockroachdb/cockroach/pkg/kv/kvserver._MultiRaft_DelegateRaftSnapshot_Handler({0x79d9920?, 0xc005a82380}, {0xb0704a8?, 0xc000f7f280})
          |   | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvserver/storage_services.pb.go:293 +0xc3
          |   | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.StreamServerInterceptor.func1({0x79d9920, 0xc005a82380}, {0xb070a90?, 0xc009c36620?}, 0xc00bfd34a0, 0x7ddc770)
          |   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:165 +0x67a
          |   | google.golang.org/grpc.getChainStreamHandler.func1({0x79d9920, 0xc005a82380}, {0xb070a90, 0xc009c36620})
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1478 +0x12c
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func4({0x79d9920, 0xc005a82380}, {0xb070a90, 0xc009c36620}, 0xc00bfd34a0, 0xc007d7cc00)
          |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:178 +0xe9
          |   | google.golang.org/grpc.getChainStreamHandler.func1({0x79d9920, 0xc005a82380}, {0xb070a90, 0xc009c36620})
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1478 +0x12c
          |   | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor({0xc0005c2000?, {{0x0?}, {0xb0750c0?, 0xc005c1e440?}}}, {0x79d9920, 0xc005a82380}, {0xb070a90, 0xc009c36620}, 0xc00bfd34a0, 0xc007d7cb80)
          |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:167 +0x517
          |   | google.golang.org/grpc.getChainStreamHandler.func1({0x79d9920, 0xc005a82380}, {0xb070a90, 0xc009c36620})
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1478 +0x12c
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0xc00628a6e0?, 0x4e057a?})
          |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:145 +0x70
          |   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc00628a6e0, {0xb03dc20, 0xc008e54390}, {0x0?, 0xc007d7cb40?}, 0xc01009b7a0)
          |   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:336 +0x148
          |   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x79d9920, 0xc005a82380}, {0xb070a90?, 0xc009c36620?}, 0xc00bfd34a0, 0xc007d7cb40)
          |   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:144 +0x14d
          |   | google.golang.org/grpc.chainStreamInterceptors.func1({0x79d9920, 0xc005a82380}, {0xb070a90, 0xc009c36620}, 0x71250c0?, 0xc000a4e1b0?)
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1469 +0xd6
          |   | google.golang.org/grpc.(*Server).processStreamingRPC(0xc0038f63c0, {0xb087d80, 0xc0084f8d00}, 0xc007039680, 0xc007199710, 0xddd71e0, 0x0)
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1627 +0x1f96
          |   | google.golang.org/grpc.(*Server).handleStream(0xc0038f63c0, {0xb087d80, 0xc0084f8d00}, 0xc007039680, 0x0)
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1708 +0xfaf
          |   | google.golang.org/grpc.(*Server).serveStreams.func1.2()
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:965 +0xed
          |   | created by google.golang.org/grpc.(*Server).serveStreams.func1
          |   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:963 +0x4de
          | Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *markers.withMark (4) *errbase.opaqueWrapper (5) *errutil.withPrefix (6) *errbase.opaqueWrapper (7) *errutil.withPrefix (8) *errbase.opaqueWrapper (9) *errutil.leafError
        Error types: (1) *barriers.barrierErr
            --- FAIL: TestTxnNegotiateAndSendDoesNotBlock/multiRange=false/strict=false/routeNearest=true (234.53s)
=== RUN   TestTxnNegotiateAndSendDoesNotBlock/multiRange=false/strict=true
        --- FAIL: TestTxnNegotiateAndSendDoesNotBlock/multiRange=false/strict=true (470.04s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTxnNegotiateAndSendDoesNotBlock.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31416

--- a/pkg/kv/kvserver/rditer/BUILD.bazel
+++ b/pkg/kv/kvserver/rditer/BUILD.bazel
@@ -15,7 +15,6 @@ go_library(
         "//pkg/roachpb",
         "//pkg/storage",
         "//pkg/storage/enginepb",
-        "//pkg/util/hlc",
         "//pkg/util/iterutil",
         "@com_github_cockroachdb_pebble//:pebble",
         "@com_github_cockroachdb_pebble//rangekey",

--- a/pkg/kv/kvserver/rditer/replica_data_iter.go
+++ b/pkg/kv/kvserver/rditer/replica_data_iter.go
@@ -17,7 +17,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage"
-	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
 	"github.com/cockroachdb/pebble"
 	"github.com/cockroachdb/pebble/rangekey"
@@ -88,16 +87,19 @@ func MakeReplicatedKeySpans(d *roachpb.RangeDescriptor) []roachpb.Span {
 // creates a SpanSet instead of a slice of spans. Note that lock table spans
 // are skipped.
 func MakeReplicatedKeySpanSet(d *roachpb.RangeDescriptor) *spanset.SpanSet {
-	spans := MakeReplicatedKeySpans(d)
-	if len(spans) != 5 {
-		panic("unexpected output from MakeReplicatedKeySpans")
-	}
+	spans := Select(d.RangeID, SelectOpts{
+		ReplicatedBySpan:    d.RSpan(),
+		ReplicatedByRangeID: true,
+		// NB: We don't need to add lock table spans. The caller is expected to add
+		// these.
+		ReplicatedSpansFilter: ReplicatedSpansExcludeLocks,
+	})
 	ss := spanset.New()
-	ss.AddNonMVCC(spanset.SpanReadWrite, spans[0])
-	ss.AddNonMVCC(spanset.SpanReadWrite, spans[1])
-	// NB: We don't need to add lock table spans. The caller is expected to add
-	// these.
-	ss.AddMVCC(spanset.SpanReadWrite, spans[4], hlc.MaxTimestamp)
+	for _, span := range spans {
+		// Declaring non-MVCC access to the MVCC user keyspan is equivalent to
+		// declaring access at all timestamps.
+		ss.AddNonMVCC(spanset.SpanReadWrite, span)
+	}
 	ss.SortAndDedup()
 	if err := ss.Validate(); err != nil {
 		panic(err)

--- a/pkg/kv/kvserver/rditer/replica_data_iter.go
+++ b/pkg/kv/kvserver/rditer/replica_data_iter.go
@@ -59,14 +59,38 @@ type ReplicaMVCCDataIterator struct {
 	err error
 }
 
-// makeAllKeySpans returns all key spans for the given Range, in
+// MakeAllKeySpans returns all key spans for the given Range, in
 // sorted order.
-func makeAllKeySpans(d *roachpb.RangeDescriptor) []roachpb.Span {
+func MakeAllKeySpans(d *roachpb.RangeDescriptor) []roachpb.Span {
 	return Select(d.RangeID, SelectOpts{
 		ReplicatedBySpan:      d.RSpan(),
 		ReplicatedByRangeID:   true,
 		UnreplicatedByRangeID: true,
 	})
+}
+
+// MakeAllKeySpanSet is similar to makeAllKeySpans, except it creates a SpanSet
+// instead of a slice of spans. Note that lock table spans are skipped.
+func MakeAllKeySpanSet(d *roachpb.RangeDescriptor) *spanset.SpanSet {
+	spans := Select(d.RangeID, SelectOpts{
+		ReplicatedBySpan:      d.RSpan(),
+		ReplicatedByRangeID:   true,
+		UnreplicatedByRangeID: true,
+		// NB: We don't need to add lock table spans. The caller is expected to add
+		// these.
+		ReplicatedSpansFilter: ReplicatedSpansExcludeLocks,
+	})
+	ss := spanset.New()
+	for _, span := range spans {
+		// Declaring non-MVCC access to the MVCC user keyspan is equivalent to
+		// declaring access at all timestamps.
+		ss.AddNonMVCC(spanset.SpanReadWrite, span)
+	}
+	ss.SortAndDedup()
+	if err := ss.Validate(); err != nil {
+		panic(err)
+	}
+	return ss
 }
 
 // MakeReplicatedKeySpans returns all key spans that are fully Raft

--- a/pkg/kv/kvserver/rditer/replica_data_iter_test.go
+++ b/pkg/kv/kvserver/rditer/replica_data_iter_test.go
@@ -464,7 +464,7 @@ func TestReplicaDataIteratorGlobalRangeKey(t *testing.T) {
 				if replicatedOnly {
 					expectedSpans = MakeReplicatedKeySpans(&desc)
 				} else {
-					expectedSpans = makeAllKeySpans(&desc)
+					expectedSpans = MakeAllKeySpans(&desc)
 				}
 
 				var actualSpans []roachpb.Span
@@ -500,7 +500,7 @@ func TestReplicaKeyRanges(t *testing.T) {
 		StartKey: roachpb.RKeyMin,
 		EndKey:   roachpb.RKeyMax,
 	}
-	checkOrdering(t, makeAllKeySpans(&desc))
+	checkOrdering(t, MakeAllKeySpans(&desc))
 	checkOrdering(t, MakeReplicatedKeySpans(&desc))
 	checkOrdering(t, makeReplicatedKeySpansExceptLockTable(&desc))
 }
@@ -554,7 +554,7 @@ func benchReplicaEngineDataIterator(b *testing.B, numRanges, numKeysPerRange, va
 
 	for _, desc := range descs {
 		var keyBuf roachpb.Key
-		keySpans := makeAllKeySpans(&desc)
+		keySpans := MakeAllKeySpans(&desc)
 		for i := 0; i < numKeysPerRange; i++ {
 			keyBuf = append(keyBuf[:0], keySpans[i%len(keySpans)].Key...)
 			keyBuf = append(keyBuf, 0, 0, 0, 0)

--- a/pkg/kv/kvserver/rditer/select.go
+++ b/pkg/kv/kvserver/rditer/select.go
@@ -26,6 +26,9 @@ const (
 	// ReplicatedSpansUserOnly includes just user keys, and no other replicated
 	// spans.
 	ReplicatedSpansUserOnly
+	// ReplicatedSpansExcludeLockTable includes all replicated spans except for the
+	// lock table.
+	ReplicatedSpansExcludeLocks
 )
 
 // SelectOpts configures which spans for a Replica to return from Select.
@@ -81,7 +84,7 @@ func Select(rangeID roachpb.RangeID, opts SelectOpts) []roachpb.Span {
 			sl = append(sl, makeRangeLocalKeySpan(in))
 
 			// Lock table.
-			{
+			if opts.ReplicatedSpansFilter != ReplicatedSpansExcludeLocks {
 				// Handle doubly-local lock table keys since range descriptor key
 				// is a range local key that can have a replicated lock acquired on it.
 				startRangeLocal, _ := keys.LockTableSingleKey(keys.MakeRangeKeyPrefix(in.Key), nil)

--- a/pkg/kv/kvserver/rditer/select_test.go
+++ b/pkg/kv/kvserver/rditer/select_test.go
@@ -69,6 +69,14 @@ func TestSelect(t *testing.T) {
 			filter: ReplicatedSpansUserOnly,
 		},
 		{
+			name: "r2_excludelocks",
+			sp: roachpb.RSpan{
+				Key:    roachpb.RKey("a"),
+				EndKey: roachpb.RKey("c"),
+			},
+			filter: ReplicatedSpansExcludeLocks,
+		},
+		{
 			name: "r3",
 			sp: roachpb.RSpan{
 				Key:    roachpb.RKey("a"),

--- a/pkg/kv/kvserver/rditer/testdata/TestSelect/r2_excludelocks
+++ b/pkg/kv/kvserver/rditer/testdata/TestSelect/r2_excludelocks
@@ -1,0 +1,18 @@
+echo
+----
+Select({ReplicatedBySpan:{a-c} ReplicatedSpansFilter:3 ReplicatedByRangeID:false UnreplicatedByRangeID:false}):
+  /Local/Range"{a"-c"}
+  {a-c}
+Select({ReplicatedBySpan:{a-c} ReplicatedSpansFilter:3 ReplicatedByRangeID:false UnreplicatedByRangeID:true}):
+  /Local/RangeID/123/{u""-v""}
+  /Local/Range"{a"-c"}
+  {a-c}
+Select({ReplicatedBySpan:{a-c} ReplicatedSpansFilter:3 ReplicatedByRangeID:true UnreplicatedByRangeID:false}):
+  /Local/RangeID/123/{r""-s""}
+  /Local/Range"{a"-c"}
+  {a-c}
+Select({ReplicatedBySpan:{a-c} ReplicatedSpansFilter:3 ReplicatedByRangeID:true UnreplicatedByRangeID:true}):
+  /Local/RangeID/123/{r""-s""}
+  /Local/RangeID/123/{u""-v""}
+  /Local/Range"{a"-c"}
+  {a-c}

--- a/pkg/kv/kvserver/replica_raftstorage.go
+++ b/pkg/kv/kvserver/replica_raftstorage.go
@@ -243,10 +243,10 @@ func (r *Replica) GetSnapshot(
 	if r.store.cfg.SharedStorageEnabled || storage.UseEFOS.Get(&r.ClusterSettings().SV) {
 		var ss *spanset.SpanSet
 		r.mu.RLock()
-		spans := rditer.MakeReplicatedKeySpans(r.mu.state.Desc)
+		spans := rditer.MakeAllKeySpans(r.mu.state.Desc) // needs unreplicated to access Raft state
 		startKey = r.mu.state.Desc.StartKey
 		if util.RaceEnabled {
-			ss = rditer.MakeReplicatedKeySpanSet(r.mu.state.Desc)
+			ss = rditer.MakeAllKeySpanSet(r.mu.state.Desc)
 			defer ss.Release()
 		}
 		r.mu.RUnlock()

